format.go

```diff
@@ -0,0 +1,55 @@
+package cid
+
+import (
+	mh "github.com/multiformats/go-multihash"
+)
+
+type Format interface {
+	Sum(data []byte) (*Cid, error)
+}
+
+type FormatV0 struct{}
+
+type FormatV1 struct {
+	Codec   uint64
+	HashFun uint64
+	HashLen int // 0 (or -1 for compatibility) means the default length
+}
+
+func PrefixToFormat(p Prefix) Format {
```

Comment:
Not sure if we really need this given that all `Prefix`s are `Format`s.

Reply:
The idea was that we might want to deprecate Prefix.Sum.  This way we we can format MhLength from being -1, which is a source of confusion.

Reply:
We can just check that on Sum (after a deprecation period). That is, I'd like to move away from having `MhLength == -1` (except for ID hashes, maybe). However, this is *only* an issue right now because the best way to create a CID is to construct a Prefix and then call Sum (and lazy programmers just set `MhLength = -1`).

Basically, `Prefix.Sum` isn't the problem (and is quite useful). It's `Prefix.MhLength == -1`.

Reply:
I would go one step further and not add an exception for ID hashes.  That is a round trip from Prefix.Sum(data) and back to Prefix should create identical results.  If you want to create an ID hash use the new FormatV1 struct.

Reply:
My point is that I really don't see why we can't just make `Prefix` implement `Format` (well, `Builder` now). Internally, it would just do *exactly* what the function above is doing and then call `Sum`. It just feels like adding this extra step isn't really necessary.